dlc-manager/src/sub_channel_manager.rs

```diff
@@ -692,80 +697,218 @@ where
 
     /// Start force closing the sub channel with given [`ChannelId`].
     pub fn force_close_sub_channel(&self, channel_id: &ChannelId) -> Result<(), Error> {
-        let (mut signed, state) = get_sub_channel_in_state!(
-            self.dlc_channel_manager,
-            *channel_id,
-            Signed,
-            None::<PublicKey>
+        let mut sub_channel = self
+            .dlc_channel_manager
+            .get_store()
+            .get_sub_channel(*channel_id)?
+            .ok_or(Error::InvalidParameters(format!(
+                "Unknown sub channel {:?}",
+                channel_id
+            )))?;
+
+        match sub_channel.state {
+            SubChannelState::Offered(_) => self.force_close_offered_channel(sub_channel)?,
+            SubChannelState::Accepted(ref a) => {
+                let commitment_transactions = a.commitment_transactions.clone();
+
+                self.force_close_with_saved_commitment(sub_channel, &commitment_transactions)?;
+            }
+            SubChannelState::Signed(_) | SubChannelState::Finalized(_) => {
+                self.force_close_signed_channel(sub_channel, None)?;
+            }
+            SubChannelState::Confirmed(ref c) => {
+                let commitment_transactions = c.commitment_transactions.clone();
+                self.force_close_with_saved_commitment(sub_channel, &commitment_transactions)?;
+            }
+            SubChannelState::CloseOffered(c) => {
+                sub_channel.state = SubChannelState::Signed(c.signed_subchannel);
+                self.force_close_signed_channel(sub_channel, None)?;
```

Comment:
Yeah it's quite different compared to #120. I'm not saying it's bulletproof, but it should be safe to rollback the state here. At worst, the closing fails, you're out of sync with your peer, but since you're trying to force close anyway it shouldn't matter I think.

Reply:
Right, the real problem about rolling back is to agree with your counterparty on doing so.